guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc

```diff
@@ -0,0 +1,188 @@
+and configure the Micronaut context to talk to them.
+
+Create a `@KafkaClient` to simplify publishing events in the test.
+
+test:ProductPriceChangesClient[app=testcontainers]
+
+callout:kafka-client.adoc[1]
+callout:kafka-topic-publisher.adoc[2]
+callout:kafka-key.adoc[3]
+
+We will use the Testcontainers library to spin up a Kafka and the MySQL database instances as Docker containers
+and configure the application to talk to them as follows:
+
+test:ProductPriceChangedEventHandlerTest[app=testcontainers]
+
+callout:micronaut-test-transactional-false[1]
+<2> Annotate the class with `@Property` to supply the driver class name configuration to the test.
+<3> We have configured the https://java.testcontainers.org/modules/databases/jdbc/[Testcontainers special JDBC URL] to spin up a MySQL container and configure it as a DataSource with Micronaut application context.
+callout:test-containers-disabled-without-docker[4]
+callout:test-instance-per-class[5]
+callout:test-property-provider[6]
+<7> We have used the Testcontainers JUnit 5 Extension annotations *@Testcontainers* and *@Container* to spin up a Kafka container and registered the *bootstrap-servers* location using *TestPropertyProvider* mechanism.
+<8> We have created a Product record in the database before sending the event.
+<9> During the test, we sent a message to the *product-price-changes* topic using *KafkaTemplate* with productCode as key and *ProductPriceChangedEvent* instance as value.
+<10> As Kafka message processing is an asynchronous process, we are using the *Awaitility* library to check whether the product price is updated in the database to the expected value or not with an interval of 3 seconds waiting up to a maximum of 10 seconds. If the message is consumed and processed within 10 seconds the test will pass, otherwise the test will fail.
+
+common:testApp.adoc[]
+
+You should see the Kafka and MySQL Docker containers are started and all tests should PASS.
+
+You can also notice that after the tests are executed the containers are stopped and removed automatically.
+
+== Testing Kafka integration with Test Resources
+
+common:test-resources.adoc[]
+
+=== Removing Testcontainers Dependencies
+
+Remove the <<testcontainers-dependencies,Testcontainers dependencies>> from your build files.
+
+=== Configure Test Resources
+
+common:install-test-resources.adoc[]
+
+=== Test Resources Kafka
+
+common:test-resources-kafka-test-or-dev.adoc[]
+
+=== Simpler Test with Test Resources
+
+Thanks to Test Resources, we can simplify the test as follows:
+
+test:ProductPriceChangedEventHandlerTest[app=testresources]
+
+callout:micronaut-test-transactional-false[1]
+<2> We have created a Product record in the database before sending the event.
+<3> During the test, we sent a message to the *product-price-changes* topic using *KafkaTemplate* with productCode as key and *ProductPriceChangedEvent* instance as value.
```

Comment:
Update `KafkaTemplate` to `ProductPriceChangesClient`.

Reply:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/1332/commits/a953961322ba0f99e4c29d78b52456a50b44fef3